x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlRealm.java

```diff
@@ -744,25 +785,69 @@ private static Tuple<AbstractReloadingMetadataResolver, Supplier<EntityDescripto
         FileWatcher watcher = new FileWatcher(path);
         watcher.addListener(new FileListener(logger, resolver::refresh));
         watcherService.add(watcher, ResourceWatcherService.Frequency.MEDIUM);
-        return new Tuple<>(resolver, () -> resolveEntityDescriptor(resolver, entityId, path.toString()));
+        return new Tuple<>(resolver, () -> resolveEntityDescriptor(resolver, entityId, path.toString(), true));
     }
 
     private static EntityDescriptor resolveEntityDescriptor(
         AbstractReloadingMetadataResolver resolver,
         String entityId,
-        String sourceLocation
+        String sourceLocation,
+        boolean failOnError
     ) {
         try {
-            final EntityDescriptor descriptor = resolver.resolveSingle(new CriteriaSet(new EntityIdCriterion(entityId)));
+            final EntityDescriptor descriptor = resolveEntityDescriptorWithPossibleRefresh(resolver, entityId);
             if (descriptor == null) {
-                throw SamlUtils.samlException("Cannot find metadata for entity [{}] in [{}]", entityId, sourceLocation);
+                if (failOnError) {
+                    throw SamlUtils.samlException("Cannot find metadata for entity [{}] in [{}]", entityId, sourceLocation);
+                } else {
+                    logger.warn(
+                        "cannot load SAML metadata for [{}] from [{}]; SAML authentication for this realm will fail",
+                        entityId,
+                        sourceLocation
+                    );
+                    return new UnresolvedEntity(entityId, sourceLocation);
+                }
             }
             return descriptor;
         } catch (ResolverException e) {
             throw SamlUtils.samlException("Cannot resolve entity metadata", e);
         }
     }
 
+    private static EntityDescriptor resolveEntityDescriptorWithPossibleRefresh(AbstractReloadingMetadataResolver resolver, String entityId)
+        throws ResolverException {
+        var criteria = new CriteriaSet(new EntityIdCriterion(entityId));
+        EntityDescriptor descriptor = resolver.resolveSingle(criteria);
+        if (descriptor == null) {
+            /*
+             * If the descriptor is null, we haven't found a metadata file with that entity id in it. This could be caused by 2 things:
+             * 1. We didn't find any metadata file (e.g. the metadata is loaded over http, and the request returned an error
+             * 2. The file does exist, but doesn't contain the entity.
+             * In either case it's worth refreshing the metadata again to see if it's now correct because it's possible that the problem
+             *    has been resolved (although if metadata is loaded from a local file we monitor it for changes anyway, so this refresh
+             *    is unlikely to have any benefit).
```

Comment:
Can we check the `resolver` object type and only perform refresh if it is a `HTTPMetadataResolver`?

Reply:
We can, do we care enough about that to make the code more complex?